### PR TITLE
Reusable workflow to build and optionally publish docker images

### PR DESCRIPTION
**Pull request by DMRobertson**
Cribbed from Synapse's config.

After some back-and-forth, I decided to always build amd64 and arm64 to keep things simple.

If I'm following the recommended branch and tag system recommended for github actions, then I think I need to

- merge this to main
- fast forward releases/v1 to the merged commit
- force push the v1 tag to the merged commit too

### CODE REVIEW

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+      - name: Build and push arm64
+        if: ${{ inputs.arm64 }}
+        uses: docker/build-push-action@v2
+        with:
+          push: ${{ inputs.publish }}
+          labels: "gitsha1=${{ github.sha }}"
+          tags: "matrixdotorg/${{ inputs.image-name }}:${{ steps.set-tag.outputs.tag }}"
+          file: ${{ inputs.dockerfile }}
+          platforms: linux/arm64
```
Comment (reivilibre) [inline .github/workflows/docker.yml]: Have you tried this? I'm not 100% sure, but I think this means the ARM64 image will overwrite the tag for the AMD64 image.

Why not accept the list of platforms as an input? TBH though, I'd be happy to always produce the ARM64 image. It's gaining popularity, it's *so easy* for us to do, but *so annoying* for users of the packages when it's missing.

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+on:
+  workflow_call:
+    inputs:
+      arm64:
+        description: If set to true, build an image for arm64 as well as amd64.
+        type: boolean
+        required: false
+        default: false
+      image-name:
+        description: "What to call the docker image. This is the FOO in 'matrixdotorg/FOO:tagname'"
```
Comment (reivilibre) [inline .github/workflows/docker.yml]: I would choose to make the end-user include `matrixdotorg/` in the name they specify.

For two reasons really;
* searchability in the repo that uses this action; and mainly
* when prototyping, it's helpful to publish to a personal namespace.

**Review by reivilibre — DISMISSED** [review]
Rest of it seems good to me

File: .github/workflows/docker.yml (hunk 2 above)
Comment (DMRobertson) [reply .github/workflows/docker.yml]: Agreed, thanks

File: .github/workflows/docker.yml (hunk 1 above)
Comment (DMRobertson) [reply .github/workflows/docker.yml]: Just caution. I don't want to give the impression that we support running all of our stuff on ARM unless we're all in agreement.

**Comment by DMRobertson** [comment]
Testing this at https://github.com/matrix-org/sydent/pull/492, which is now unhappy on the arm build.

File: .github/workflows/docker.yml (hunk 1 above)
Comment (DMRobertson) [reply .github/workflows/docker.yml]: > Have you tried this?

Thanks for the pointer to allow a custom name. It works: https://hub.docker.com/r/dmrobertsonelement/sydent

(though it'd be snappier if there were more arm wheels)

**Comment by DMRobertson** [comment]
I think @reivilibre is out today. Would you mind taking a quick look at this @squahtx?

Edit: (given the context on the sydent issue above)

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        uses: docker/setup-buildx-action@v1
+
+      - name: Inspect builder
+        run: docker buildx inspect
+          
+      - name: Log in to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
Comment (squahtx) [inline .github/workflows/docker.yml]:
```suggestion
          username: ${{ secrets.DOCKER_HUB_USERNAME }}
          password: ${{ secrets.DOCKER_HUB_TOKEN }}
```

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up QEMU
+        if: contains(inputs.platforms, 'linux/arm64')
+        uses: docker/setup-qemu-action@v1
+        with:
+          platforms: arm64
```
Comment (squahtx) [inline .github/workflows/docker.yml]: Looks questionable, given that additional platforms may be specified in `inputs.platforms`.  
Unless the intent is for platforms to be a subset of { linux/amd64, linux/arm64 }?

File: .github/workflows/docker.yml (hunk 3 above)
Comment (squahtx) [reply .github/workflows/docker.yml]: Should these be updated?

**Comment by squahtx** [comment]
Looks good minus two questions.
I only reviewed from @reivilibre's review onwards.

File: .github/workflows/docker.yml (hunk 3 above)
Comment (DMRobertson) [reply .github/workflows/docker.yml]: Argh, yes---thank you.

File: .github/workflows/docker.yml (hunk 4 above)
Comment (DMRobertson) [reply .github/workflows/docker.yml]: Ahh, I hadn't grokked that there are other platforms available. https://github.com/docker/setup-qemu-action lists e.g. RISCV, older ARMs, MIPS...

I don't think we're planning to use any of these apart from arm64 imminently, but we may as well keep the door open. I guess I could change the condition to setup qemu if and only if `inputs.platforms != 'linux/amd64'`?

Comment (squahtx) [reply .github/workflows/docker.yml]: The
```
        with:
          platforms: arm64
```
bit is going to be troublesome IMO. Unless we can do some magic?

Comment (DMRobertson) [reply .github/workflows/docker.yml]: Maybe it's just easier to say we only support amd64 and arm64. If we want something more complicated we can revisit this later.

Comment (squahtx) [reply .github/workflows/docker.yml]: Sounds good. Let's make a note of it in the description for `platforms`?

Comment (DMRobertson) [reply .github/workflows/docker.yml]: On third thought, maybe it's easier to just keep it simple and always build amd64+arm64 and not provide the option to customise this. [Olivier was happy with this](https://github.com/matrix-org/backend-meta/pull/2#discussion_r804602636)... and if it's good enough for Synapse, it should be good enough for all our projects.

**Review by squahtx — APPROVED** [review]

**Pull request merged**